webpack/assets/javascripts/react_app/components/HostDetails/Properties/OSIcon.js

```diff
@@ -0,0 +1,82 @@
+import PropTypes from 'prop-types';
+import React, { useCallback, useState, useEffect } from 'react';
+
+const OsIcon = ({ os, family }) => {
+  const [image, setImage] = useState(null);
+  const fileName = useCallback(() => {
+    switch (true) {
+      case /fedora/i.test(os):
+        return 'fedora';
+      case /ubuntu/i.test(os):
+        return 'ubuntu';
+      case /solaris|sunos/i.test(os):
+        return 'stub/steelblue-s';
+      case /darwin/i.test(os):
+        return 'stub/darkred-d';
+      case /centos/i.test(os):
+        return 'centos';
+      case /scientific/i.test(os):
+        return 'scientific';
+      case /archlinux/i.test(os):
+        return 'archlinux';
+      case /altlinux/i.test(os):
+        return 'stub/goldenrod-a';
+      case /gentoo/i.test(os):
+        return 'gentoo';
+      case /SLC/i.test(os):
+        return 'stub/blue-s';
+      case /FreeBSD/i.test(os):
+        return 'freebsd';
+      case /aix/i.test(os):
+        return 'stub/forestgreen-a"';
+      case /VRP/i.test(os):
+        return 'stub/firebrick-h';
+      case /Junos/i.test(os):
+        return 'stub/darkblue-j';
+      case /OracleLinux/i.test(os):
+        return 'stub/firebrick-o';
+      case /CoreOS|ContainerLinux|Container Linux/i.test(os):
+        return 'coreos';
+      case /Flatcar/i.test(os):
+        return 'stub/darkblue-f';
+      case /RancherOS/i.test(os):
+        return 'rancheros';
+      case /NXOS/i.test(os):
+        return 'stub/darkslateblue-n';
+      case /XenServer/i.test(os):
+        return 'stub/black-x';
+      case /Puppet/i.test(os):
+        return 'stub/goldenrod-p';
+      case /Windows/i.test(os):
+        return 'stub/steelblue-w';
+      default:
+        if (!family || !family.length) return 'stub/black-x';
+        return family.toLowerCase();
+    }
+  }, [os, family]);
+
+  useEffect(() => {
+    const file = fileName();
+    const loadImage = async () => {
+      const loaded = await import(
+        `../../../../../images/icons16x16/${file}.png`
+      );
+      setImage(loaded.default);
+    };
+    loadImage();
+  }, [os, fileName]);
```

Comment:
I'm curious, why do we need to async load the images using js (and duplicate all image files for it) instead of just pointing the `img src` to the correct path in the public asset folder?

Reply:
doesn't the icon path change due to the rails' assets pipeline?
for example, this is the path for the fedora icon - `/assets/icons16x16/fedora-55952132f20fa7c3395608ad4972cf2f45e284e5bf3d13f7b31d1a0b3575b462.png`

Reply:
They are indeed served with unique names (hashes). This allows us to send expire headers allowing clients to cache these files forever. This makes pages faster to load. Remember that at least on EL7 there's no HTTP2 so a maximum number of concurrent HTTP requests. It would be nice if we could maintain this browser side caching of assets. AFAIK we also send these headers for webpack directories in a production environment.

I can't say exactly from this code, but is import on the client side or server side? The browser should **never** get `/assets/images/icons/16x16/${file}.png` because the client may cache it in Foreman 2.1. If the image changes (like when the RH logo changed), they will not get the new version in 2.2.

Reply:
hmm, IIUC, all of the images are compiled into the bundle now and served with it, and when the OS is loaded the specific image is loaded into the img tag as base64. 
Considering there are ~300 different letter and color combinations, I don't think that we should include them all in the bundle.
In fact. there are only 12 actual logos there, while the rest are just letters in colored circles. 
Would it be better to have a js helper that takes a letter and color and just creates a colorful circle around it? It could be useful for other places as well.
Maybe all the 12 actual logos should actually live in /public. Other than the one time RH changed its logo, I don't think we've ever had any of them change, and for the one time in a decade a logo changes we can probably handle it with a new file name.